src/ui/FunctionsOutlineView.cpp

```diff
@@ -0,0 +1,295 @@
+#include <LayoutBuilder.h>
+#include <NaturalCompare.h>
+#include <OutlineListView.h>
+#include <ScrollView.h>
+#include <StringView.h>
+#include <Window.h>
+
+#include "Editor.h"
+#include "EditorTabManager.h"
+#include "GenioWindow.h"
+#include "GenioWindowMessages.h"
+#include "LSPEditorWrapper.h"
+#include "ToolBar.h"
+
+#include "Log.h"
+
+#define kGoToSymbol	'gots'
+#define kMsgSort 'sort'
+#define kMsgCollapseAll 'coll'
+
+static bool sSorted = false;
+static bool sCollapsed = false;
+
+class SymbolListItem : public BStringItem {
+public:
+		SymbolListItem(BMessage& details)
+			: BStringItem(details.GetString("name"))
+			, fDetails ( details )
+			{}
+
+		const BMessage& Details() const { return fDetails; }
+private:
+		BMessage	fDetails;
+};
+
+
+static int
+CompareItems(const BListItem* itemA, const BListItem* itemB)
+{
+	return itemA - itemB;
+}
+
+
+static int
+CompareItemsText(const BListItem* itemA, const BListItem* itemB)
+{
+	return BPrivate::NaturalCompare(((BStringItem*)itemA)->Text(), ((BStringItem*)itemB)->Text());
+}
+
+
+#undef B_TRANSLATION_CONTEXT
+#define B_TRANSLATION_CONTEXT "FunctionsOutlineView"
+
+
+FunctionsOutlineView::FunctionsOutlineView()
+	:
+	BView(B_TRANSLATE("Class outline"), B_WILL_DRAW),
+	fListView(nullptr),
+	fToolBar(nullptr),
+	fStatus(STATUS_EMPTY),
+	fSymbolsLastUpdateTime(0)
+{
+	SetFlags(Flags() | B_PULSE_NEEDED);
+
+	fListView = new BOutlineListView("listview");
+	fScrollView = new BScrollView("scrollview", fListView,
+		B_FRAME_EVENTS | B_WILL_DRAW, true, true, B_FANCY_BORDER);
+	fToolBar = new ToolBar();
+	fToolBar->ChangeIconSize(16);
+	// TODO: Icons
+	//fToolBar->AddAction(kMsgCollapseAll, "Collapse all", "kIconGitRepo", true);
+	fToolBar->AddAction(kMsgSort, B_TRANSLATE("Sort"), "kIconOutlineSort", true);
+	fToolBar->SetExplicitMinSize(BSize(250, B_SIZE_UNSET));
+	fToolBar->SetExplicitMaxSize(BSize(B_SIZE_UNLIMITED, B_SIZE_UNSET));
+	BLayoutBuilder::Group<>(this)
+		.AddGroup(B_VERTICAL, 0)
+			.Add(fToolBar)
+			.Add(fScrollView)
+		.End();
+
+	_UpdateDocumentSymbols(nullptr);
+}
+
+
+/* virtual */
+void
+FunctionsOutlineView::AttachedToWindow()
+{
+	fToolBar->SetTarget(this);
+	if (LockLooper()) {
+		Window()->StartWatching(this, MSG_NOTIFY_EDITOR_SYMBOLS_UPDATED);
+		UnlockLooper();
+	}
+}
+
+
+/* virtual */
+void
+FunctionsOutlineView::DetachedFromWindow()
+{
+	if (LockLooper()) {
+		Window()->StopWatching(this, MSG_NOTIFY_EDITOR_SYMBOLS_UPDATED);
+		UnlockLooper();
+	}
+}
+
+
+/* virtual */
+void
+FunctionsOutlineView::Pulse()
+{
+	Editor* editor = gMainWindow->TabManager()->SelectedEditor();
+	// We already requested symbols, just need to wait a bit
+	if (editor != nullptr && *editor->FileRef() == fCurrentRef &&
+			fStatus == STATUS_PENDING)
+		return;
+
+	// TODO: Not very nice: we should request symbols update only if there's a
+	// change. But we shouldn't ask one for EVERY change.
+	// Maybe editor knows, but I wasn't able to make it work there
+	// Update every 10 seconds
```

Comment:
Why don't we set a timer after being idle for a while? This could be done with BMessageRunner, I think.

Reply:
Good idea. I'll try

Reply:
I have tried briefly, but it doesn't change much: in the end it's only a style preference